### PR TITLE
Enable object and array type parameters for template deployment

### PR DESCRIPTION
**Pull request by TianoMS**
Issue: https://github.com/Azure/azure-powershell/issues/1940

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@TianoMS__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by TianoMS** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/771/

File: src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/GalleryTemplatesClient.cs [hunk 1]
```diff
@@ -401,6 +404,15 @@ private Type GetParameterType(string resourceParameterType)
...
                 typeObject = typeof(bool);
             }
+            else if (resourceParameterType.Equals(objectType, StringComparison.OrdinalIgnoreCase)
+                || resourceParameterType.Equals(secureObjectType, StringComparison.OrdinalIgnoreCase))
+            {
+                typeObject = typeof(Hashtable);
+            }
+            else if (resourceParameterType.Equals(arrayType, StringComparison.OrdinalIgnoreCase))
+            {
+                typeObject = typeof(object[]);
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/GalleryTemplatesClient.cs]: Hashtable[] doesn't work?

Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/GalleryTemplatesClient.cs]: For Json "array" type, we can't map it to Hashtable[]. Here we are actually binding to a dynamic parameter in powershell. So it's for something like "@('abc', 'def')", which is maps to object[].

File: src/ResourceManager/Resources/Commands.Resources.Test/Resources/sampleTemplateParameterFile.json [hunk 2]
```diff
@@ -8,7 +8,22 @@
     "string" : {
         "value" : "myvalue"
     },
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.Resources.Test/Resources/sampleTemplateParameterFile.json]: Excellent! Thanks for adding these to our existing test.

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/DeploymentTests.ps1 [hunk 3]
```diff
@@ -219,3 +219,135 @@ function Test-NewDeploymentWithKeyVaultReference
...
+	$rgname = Get-ResourceGroupName
+	$rname = Get-ResourceName
+	$rglocation = "EastUS"
+
+	try
+	{
+		# Test
+		New-AzureRmResourceGroup -Name $rgname -Location $rglocation
+		
+		$deployment = New-AzureRmResourceGroupDeployment -Name $rname -ResourceGroupName $rgname -TemplateFile complexParametersTemplate.json -TemplateParameterObject @{appSku=@{code="f1"; name="Free"}; servicePlan="plan1"; ranks=@("c", "d")}
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/DeploymentTests.ps1]: Nice!

**Comment by hovsepm** [comment]
@TianoMS please pull from upstream and resolve conflicts.

**Comment by TianoMS** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/778/

**Pull request merged**